mlflow/server/jobs/util.py

```diff
@@ -121,6 +146,211 @@ def is_process_alive(pid: int) -> bool:
+        job_store.mark_job_timed_out(job_id)
+
+
+@dataclass
+class HueyInstance:
+    instance: Any
+    submit_task: Callable[..., Any]
+
+
+_huey_instance_map: dict[str, HueyInstance] = {}
+_huey_instance_map_lock = threading.RLock()
+
+
+def _get_or_init_huey_instance(instance_key: str):
+    from huey import SqliteHuey
+    from huey.serializer import Serializer
+
+    class CloudPickleSerializer(Serializer):
+        def serialize(self, data):
+            return cloudpickle.dumps(data)
+
+        def deserialize(self, data):
+            return cloudpickle.loads(data)
+
+    with _huey_instance_map_lock:
+        if instance_key not in _huey_instance_map:
+            _logger.info(f"Creating huey instance for {instance_key}")
+            huey_store_file = os.path.join(
+                os.environ[HUEY_STORAGE_PATH_ENV_VAR], f"{instance_key}.mlflow-huey-store"
+            )
+            huey_instance = SqliteHuey(
+                filename=huey_store_file,
+                results=False,
+                serializer=CloudPickleSerializer(),
+            )
+            huey_submit_task_fn = huey_instance.task(retries=0)(_exec_job)
+            _huey_instance_map[instance_key] = HueyInstance(
+                instance=huey_instance,
+                submit_task=huey_submit_task_fn,
+            )
+        return _huey_instance_map[instance_key]
+
+
+def _launch_huey_consumer(
+    job_fn_fullname: str,
+) -> None:
+    _logger.info(f"Starting huey consumer for job function {job_fn_fullname}")
+    job_fn = _load_function(job_fn_fullname)
+
+    if not hasattr(job_fn, "_job_fn_metadata"):
+        raise MlflowException(
+            f"The job function {job_fn_fullname} is not decorated by "
+            "'mlflow.server.jobs.job_function'."
+        )
+
+    max_job_parallelism = job_fn._job_fn_metadata.max_workers
+
+    def _huey_consumer_thread() -> None:
+        while True:
+            # start MLflow job runner process
+            # Put it inside the loop to ensure the job runner process alive
+            job_runner_proc = _start_huey_consumer_proc(
+                job_fn_fullname,
+                max_job_parallelism,
+            )
+            job_runner_proc.wait()
+            time.sleep(1)
+
+    # start job runner.
+    threading.Thread(
+        target=_huey_consumer_thread,
+        name=f"MLflow-huey-consumer-{job_fn_fullname}-watcher",
+        daemon=True,
+    ).start()
+
+
+def _launch_job_runner(backend_store_uri, env_map, server_proc_pid):
+    from mlflow.exceptions import MlflowException
+    from mlflow.utils.process import _exec_cmd
+    from mlflow.utils.uri import extract_db_type_from_uri
+
+    try:
+        extract_db_type_from_uri(backend_store_uri)
+    except MlflowException:
+        _logger.warning(
+            f"Job store requires a database backend store URI but got {backend_store_uri}, "
+            "skip launching the job runner."
+        )
+        return
+
+    return _exec_cmd(
```

Comment:
Is there a reason _exec_cmd is used? Can we use `subprocess.Popen`?

Reply:
We can use `subprocess.Popen`

Reply:
updated code to use `subprocess.Popen` instead.